ftrace_hook.c

```diff
@@ -92,15 +102,15 @@ static int fh_resolve_hook_address(struct ftrace_hook *hook)
 }
 
 static void notrace fh_ftrace_thunk(unsigned long ip, unsigned long parent_ip,
-		struct ftrace_ops *ops, struct pt_regs *regs)
+		struct ftrace_ops *ops, struct ftrace_regs *regs)
 {
 	struct ftrace_hook *hook = container_of(ops, struct ftrace_hook, ops);
 
 #if USE_FENTRY_OFFSET
-	regs->ip = (unsigned long) hook->function;
+	regs->regs.ip = (unsigned long)hook->function;
```

Comment:
Initially I was a bit confused by this, but it's actually a pretty clever hack for older kernels, since `struct pt_regs` and the `struct ftrace_regs` defined above have the same memory layout, this assignment works as expected.

Reply:
(Though, from what I can tell from LKML, the *proper* way to get access to `struct pt_regs` in `struct ftrace_regs` seems to be `ftrace_get_regs()`...)

Reply:
Maybe something like this would be a “better-looking” portability shim:

```c
#if LINUX_VERSION_CODE < KERNEL_VERSION(5,11,0)
struct ftrace_regs {
	struct pt_regs regs;
};

static __always_inline struct pt_regs *ftrace_get_regs(struct ftrace_regs *fregs)
{
	return &fregs->regs;
}
#endif
```
```c
static void notrace fh_ftrace_thunk(unsigned long ip, unsigned long parent_ip,
		struct ftrace_ops *ops, struct ftrace_regs *fregs)
{
	struct pt_regs *regs = ftrace_get_regs(fregs);
	// ...
}
```